src/Http/Server.php

```diff
@@ -84,59 +93,70 @@ public static function get($key = null, $default = null)
     }
 
     /**
-     * Help to sanitize some _SERVER keys
+     * Returns the correct host
      *
-     * @param string $key
-     * @param mixed $value
-     * @return mixed
+     * @param bool $forwarded Deprecated. Todo: remove in 3.7.0
+     * @return string
      */
-    public static function sanitize(string $key, $value)
+    public static function host(bool $forwarded = false): string
     {
-        // make sure $value is not null
-        $value ??= '';
+        $host = null;
 
-        switch ($key) {
-            case 'SERVER_ADDR':
-            case 'SERVER_NAME':
-            case 'HTTP_HOST':
-            case 'HTTP_X_FORWARDED_HOST':
-                $value = strip_tags($value);
-                $value = preg_replace('![^\w.:-]+!iu', '', $value);
-                $value = trim($value, '-');
-                $value = htmlspecialchars($value, ENT_COMPAT);
-                break;
-            case 'SERVER_PORT':
-            case 'HTTP_X_FORWARDED_PORT':
-                $value = (int)(preg_replace('![^0-9]+!', '', $value));
-                break;
+        // insecure host parameters are only allowed when hosts
+        // are validated against set of host patterns
+        $allowInsecure = empty(static::$hosts) === false;
+
+        if ($allowInsecure === true && static::isBehindProxy() === true) {
+            $host = static::get('HTTP_X_FORWARDED_HOST');
         }
 
-        return $value;
+        if (empty($host) === true) {
+            $host = static::get('SERVER_NAME');
+        }
+
+        if ($allowInsecure === true && empty($host) === true) {
+            $host = static::get('HTTP_HOST');
+        }
+
+        if (empty($host) === true) {
+            $host = static::get('SERVER_ADDR');
+        }
+
+        // ignore invalid host names
+        if (static::isAllowedHost($host) === false) {
+            $host = null;
```

Comment:
@distantnative Yes, I think it should throw exception from here as we discussed in https://github.com/getkirby/kirby/pull/4173/files#r812212533